dll/win32/shell32/shlexec.cpp

```diff
@@ -458,6 +458,144 @@ static HRESULT SHELL_GetPathFromIDListForExecuteW(LPCITEMIDLIST pidl, LPWSTR psz
+    HWND progman = GetShellWindow();
+    if (hwnd && IsWindowVisible(hwnd) && hwnd != GetDesktopWindow() &&
+        hwnd != progman && !IsChild(progman, hwnd))
+    {
+        return hwnd;
+    }
+    return NULL;
+}
+
+/*************************************************************************
+ *    SHELL_PromptAndRunProcessAs [Internal]
+ */
+typedef struct {
+    WCHAR NameBuffer[MAX_PATH], Pass[MAX_PATH];
+    LPWSTR Name, Domain;
+    BOOL Safer;
+} RUNASDLGDATA;
+
+static
+INT_PTR
+CALLBACK
+RunAsDlgProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    RUNASDLGDATA *pData = (RUNASDLGDATA *)GetWindowLongPtrW(hwnd, DWLP_USER);
+    switch (uMsg)
+    {
+        case WM_INITDIALOG:
+        {
+            SetWindowLongPtrW(hwnd, DWLP_USER, (LPARAM)(pData = (RUNASDLGDATA *)lParam));
+            SendDlgItemMessageW(hwnd, IDC_RUNAS_NAME, EM_LIMITTEXT, _countof(RUNASDLGDATA::NameBuffer)-1, 0);
+            SendDlgItemMessageW(hwnd, IDC_RUNAS_PASS, EM_LIMITTEXT, _countof(RUNASDLGDATA::Pass)-1, 0);
+            SendDlgItemMessageW(hwnd, IDC_RUNAS_OTHER, BM_CLICK, 0, 0);
+
+            HWND control = GetDlgItem(hwnd, IDC_RUNAS_THIS);
+            WCHAR fmtbuf[200], buf[_countof(fmtbuf) + _countof(RUNASDLGDATA::NameBuffer)];
+            DWORD cch = _countof(RUNASDLGDATA::NameBuffer);
+            if (GetUserNameW(pData->NameBuffer, &cch))
+            {
+                SendMessageW(control, WM_GETTEXT, _countof(fmtbuf), (LPARAM)buf);
+                StringCchPrintfW(fmtbuf, _countof(fmtbuf), buf, L"(%s)"); // Change "Blah blah %s" to "Blah blah (%s)"
+                StringCchPrintfW(buf, _countof(buf), fmtbuf, pData->NameBuffer);
+                SendMessageW(control, WM_SETTEXT, 0, (LPARAM)buf);
+                SendDlgItemMessageW(hwnd, IDC_RUNAS_NAME, CB_ADDSTRING, 0, (LPARAM)pData->NameBuffer);
+            }
+            SendDlgItemMessageW(hwnd, IDC_RUNAS_NAME, CB_SETCURSEL, 0, 0);
+            return TRUE;
+        }
+
+        case WM_COMMAND:
+            switch (LOWORD(wParam))
+            {
+                case IDCANCEL:
+                    EndDialog(hwnd, 1);
+                    break;
+
+                case IDOK:
+                {
+                    pData->Safer = SendDlgItemMessageW(hwnd, IDC_RUNAS_SAFER,
+                                                      BM_GETCHECK, 0, 0);
+                    SendDlgItemMessageW(hwnd, IDC_RUNAS_NAME, WM_GETTEXT,
+                                       _countof(RUNASDLGDATA::NameBuffer), (LPARAM)pData->NameBuffer);
+                    SendDlgItemMessageW(hwnd, IDC_RUNAS_PASS, WM_GETTEXT,
+                                       _countof(RUNASDLGDATA::Pass), (LPARAM)pData->Pass);
+                    UINT other = SendDlgItemMessageW(hwnd, IDC_RUNAS_OTHER,
+                                                    BM_GETCHECK, 0, 0) != 0;
+                    pData->Name = pData->NameBuffer;
+                    pData->Domain = strstrW(pData->Name, L"\\");
```

Comment:
```suggestion
                    pData->Domain = wcsstr(pData->Name, L"\\");
```
instead of using that wine-specific function.